metagraph/src/cli/transform_anno_tax.cpp

```diff
@@ -0,0 +1,111 @@
+#include "transform_anno_tax.hpp"
+
+#include <filesystem>
+#include <tsl/hopscotch_set.h>
+
+#include "annotation/representation/annotation_matrix/annotation_matrix.hpp"
+#include "annotation/taxonomy/taxonomic_db.hpp"
+#include "common/threads/threading.hpp"
+#include "config/config.hpp"
+#include "cli/load/load_annotation.hpp"
+
+#include "common/logger.hpp"
+
+
+namespace mtg {
+namespace cli {
+
+using mtg::common::logger;
+
+std::vector<std::string> get_anno_filenames(const std::vector<std::string> &paths) {
+    std::vector<std::string> filenames;
+
+    for (const std::string &path: paths) {
+        if (! filesystem::is_directory(path)) {
+            if (!utils::ends_with(path, ".annodbg")) {
+                logger->warn("File '{}' is not an '.annodbg' file. No evaluation for this file.", path);
+                continue;
+            }
+            filenames.push_back(path);
+            continue;
+        }
+        logger->trace("Looking for anno files in directory: '{}'", path);
+        for (auto &rec_file: filesystem::recursive_directory_iterator(path)) {
+            if (utils::ends_with(rec_file.path().string(), ".annodbg")) {
+                logger->trace("Found anno file '{}'", rec_file.path().string());
+                filenames.push_back(rec_file.path().string());
+            }
+        }
+    }
+    return filenames;
+}
+
+void call_taxonomy_map_updates(annot::TaxonomyDB &taxonomy,
+                               const std::vector<std::string> &filenames,
+                               cli::Config *config) {
+    for (uint64_t i = 0; i < filenames.size(); ++i) {
+        const auto &file = filenames[i];
+        std::unique_ptr<annot::MultiLabelEncoded<std::string>> annot =
+                cli::initialize_annotation(file, *config);
+        if (!annot->load(file)) {
+            logger->error("Cannot load annotations from file '{}'", file);
+            exit(1);
+        }
+        taxonomy.kmer_to_taxid_map_update(*annot);
+    }
```

Comment:
Make it parallel. Otherwise, there is no point to load columns in batches, if you process them sequentially anyway.

Reply:
BTW, I think it makes sense to move this code to src/annotation/taxonomy..., it's not about CLI.

Reply:
I have just realized that the parallelization proposed for taxoDB was broken. What the approach is doing is to receive pairs of {kmer, taxid} (from the annotation matrix) and update the LCA of this received 'kmer' (`taxonomic_map[kmer]`) to incorporate the 'taxid' (`taxonomic_map[kmer] = lca(taxonomic_map[kmer], taxid)`). 

The current implementation computes the `lca(taxonomic_map[kmer], taxid)` in parallel and then uses a mutex just before saving the updates in `taxonomic_map`. But this is clearly wrong because `taxonomic_map[kmer]` can be updated by an another thread while we are inside `lca(taxonomic_map[kmer], taxid)`.

So, I am not so sure how can we do anything in parallel for this taxoDB step, maybe only if we can split the anno matrix columns and make sure that each kmer is computed on one single thread. Will try to think further to see I can find any solution for this.

For your second comment, I let this `execute_fasta_file` function in CLI because it must receive `Config &config` as argument. If there is no problem to send this Config object to `src/annotation/taxonomy`, I can easily move it. The only thing was that I saw that you have never sent Config outside of CLI functions. Otherwise I don't know any other methods for getting the annotation object excepting `cli::initialize_annotation(file, *config);`

Reply:
> For your second comment, I let this execute_fasta_file function in CLI because it must receive Config &config as argument. If there is no problem to send this Config object to src/annotation/taxonomy, I can easily move it. The only thing was that I saw that you have never sent Config outside of CLI functions. Otherwise I don't know any other methods for getting the annotation object excepting cli::initialize_annotation(file, *config);

What about passing only the required parameters from config to that function, so you don't need to import cli in the core part?

Reply:
Can you please explain in more details what exactly do you mean?

Reply:
As we have discussed in the last meeting, I have deleted the parallelization for `transform_anno_tax`, because it is already fast in comparison with `tax_class`. I will come back to optimize this procedure lated, the next days I will focus on `tax_class`.

Also, I tried to minimize the piece of code here and move to the core `src/annotation/taxonomy` as soon as I finalize with constructing the annotation matrix object.